### ISSUE TITLE
Need to migrate from travis-ci.org

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#5061]
Seems like https://travis-ci.org is closing down and we have to move to https://travis-ci.com by the end of the year:
https://mailchi.mp/3d439eeb1098/travis-ciorg-is-moving-to-travis-cicom

Migration guide: https://docs.travis-ci.com/user/migrate/open-source-repository-migration

I am already experiencing a lack of responsivity of travis-ci.org for my builds: they queue for long times.
It could be that we have to look for alternatives to travis. See e.g. https://www.theregister.com/2020/11/02/travis_ci_pricng/
> “We will upgrade you to our trial (free) plan with a 10K credit allotment (which allows around 1,000 minutes in a Linux environment,” the company said. “When your credit allotment runs out - we’d love for you to consider which of our plans will meet your needs.”

Is this per 1000 min per month or total?  I suppose the latter.  It seems to get more one has to file an application. See https://travis-ci.com/plans :
> Free for open source. We love the Open Source Community, and to show how much we love it, upon validated request placed with our Support Team you may receive free OSS credits for your public builds. 

Continuing https://www.theregister.com/2020/11/02/travis_ci_pricng/ :
> There are other options for developers, including GitHub which offers 2,000 free minutes of its Actions service per month for private repositories, and unlimited free minutes for public repositories. GitLab offers 400 CI/CD minutes free for any team, and also offers unlimited free access for “qualifying open source projects.”

So maybe we can move everything to GitHub actions instead of migrating to travis-ci.com?!

**Comment by UlfNorell** [comment]
I would be for migrating to GitHub actions, but I am by no means an expert on these things. Are there things we are doing on Travis now that would be hard/impossible with GitHub actions, or is it just a matter of skilled manual labour converting the travis file?

**Comment by L-TChen** [comment]
> it just a matter of skilled manual labour converting the travis file?

As well as a vast amount of time working in the dark. 😭 

I thought that was more or less an easy task, but ... anyway see my PR #5108.

**Comment by andreasabel** [comment]
_While you are doing the hard work, @L-TChen , I wrote an email to Travis CI begging for free CI.  This might buy us (and agda/agda-stdlib etc.) some time..._

Dear Madams and Sirs of travis-ci.com,

we, the developers of the Agda programming language, have benefited from the free CI service at travis-ci.org for several years, with close to 9000 builds of

  https://travis-ci.org/github/agda/agda

and some other projects of the Agda github community, see

  https://github.com/agda/

Thanks for the great support over the years!

We would like to ask for continued support of Travis CI in form of free OSS builds on travis-ci.com.
(We are aware of the standard Free account with 10000 credits, but these would likely be consumed in just a couple builds/days.)

Please let us know whether you are willing to continue sponsorship of Agda in the form of free CI services.  Please also let us know the conditions under which such a sponsorship could take place (like e.g. our public acknowledgement of your sponsorship).  Please also be aware that Agda is a purely academic project and we have no funds for a paid subscription.

Best regards and happy holidays,
Andreas Abel  (on behalf of the Agda development team)

P.S.:  For further information on Agda see

- https://hackage.haskell.org/package/Agda
- http://agda.readthedocs.io/
- https://wiki.portal.chalmers.se/agda/pmwiki.php

and some evidence of its academic impact:

- https://wiki.portal.chalmers.se/agda/Main/PapersUsingAgda

**Comment by L-TChen** [comment]
Regarding migrating to GitHub actions, there is one caveat. Although the GitHub-hosted runner does provide Ubuntu 16.04, the pre-installed GHC is 8.10.2 ([Image Version: 20201210.0](https://github.com/actions/virtual-environments/blob/releases/ubuntu16/20201210/images/linux/Ubuntu1604-README.md)). The action  [haskell/actions/setup](https://github.com/haskell/actions/tree/main/setup) used to set up specified GHC versions only supports `cabal` from 2.2. I guess one can still manage to get older cabal manually without using the action, but it will be hard to maintain.

In short, if we move the entire CI infrastructure to GitHub Actions, we may need to give up testing some older cabal (1.24.2 and 2.0).

**Comment by andreasabel** [comment]
_I got the following reply:_

John Dimoulakis (Travis CI)                                                                                      Dec 23, 2020, 4:45 EST                                                

Hello Andreas​ ,

Thanks for contacting Travis-CI Support! We would love to help. We  offer an Open Source Subscription for free to non-commercial  open-source projects. To qualify for an Open Source subscription, the  project must meet the following requirements:
- [x] You are a project lead or regular committer (latest commit in the last month)
- [x] Project must be at least 3 months old and is in active development (with regular commits and activity)
- [x] Project meets the OSD specification
- [ ] Project must not be sponsored by a commercial company or organization (monetary or with employees paid to work on the project)
- [x] Project can not provide commercial services or distribute paid versions of the software
 
Does this sound like you and your project? We'd be very happy to support you! However, if your project does not match these requirements or you have further questions [1], please feel free to ask!

We look forward to your response if you meet these requirements to proceed with the next steps.

Thank you

[1] https://docs.travis-ci.com/user/billing-faq/

_The only unsure point is 4, whether our Chalmers/GU salaries count towards sponsoring by a commercial company or organization._

**Comment by andreasabel** [comment]
_I answered:_

Hello John,

thanks for your quick reply.

We meet the requirements for an OSS, the only clarification I need is on point 4:

>   * Project must not be sponsored by a commercial company or organization (monetary or with employees paid to work on the project)

We are academics (e.g., I am a senior lecturer) and get a salary from the university, and part of our time goes into Agda development.  I would guess that universities do not count as "commercial company or organization", do they?

My other questions concerns our development as a GitHub organization ("agda").  Should then

1. all of its members apply for an OSS individually, or
2. do we get an OSS as a organization, or
3. should we select one member that holds the OSS for the organization?

Best regards and happy holidays,
Andreas

**Comment by UlfNorell** [comment]
We dropped travis now didn't we? Last build was 11 days ago. I don't see any strong reasons to go back even if we get OSS status, so maybe close this for now?

**Comment by asr** [comment]
> Last build was 11 days ago.

Because you need to include the sentence 'complete tests' in the commit message for running the Travis tests.

At least the `hTags`test is missing in the GitHub workflows.

**Comment by L-TChen** [comment]
No, there are still some leftovers on Travis. And I don’t see an easy way to port them to GitHub right now.

**Comment by jespercockx** [comment]
Is this something that needs to be fixed before the release of 2.6.2?

**Comment by UlfNorell** [comment]
Interaction tests are currently not working (ping @L-TChen )
```
Issue983.hs:«line»:«col»: error:
Could not load module ‘Agda.Version’
It is a member of the package ‘Agda-2.6.2-AltzoNQZhwF6I1cCJZseYT’
which is unusable due to missing dependencies:
aeson-1.5.5.1 async-2.2.2 bifunctors-5.5.10 ...
```
This is a test case using Agda as a Haskell library (there are 15 such tests in `test/interaction`). Looks like new-cabal nonsense and we need to have a cabal file for the interaction tests and use `cabal runghc` or some such.

**Comment by L-TChen** [comment]
The following message is copied from #5164.

> I am not so sure why stack may have a different build plan even for the same `stack.yaml`, so the cached build plan is different from the build plan supposed to use. I will use the [lock file](https://docs.haskellstack.org/en/stable/lock_files/) instead as the cache key. It seems fine on my fork https://github.com/L-TChen/agda/runs/1813818417?check_suite_focus=true. I will push this tomorrow.

**Comment by andreasabel** [comment]
Short of 4 months later travis-ci.com comes back with the following offer:
> Apr 19, 2021, 17:39 EDT
> 
> Hi Andreas,
> 
> Please accept our sincere apologies for any inconvenience caused due to 
> delays in processing your open-source credit request while we made 
> improvements to the platform.
> 
> Open source has always been and always will be at the core of what 
> Travis CI stands for. We are pleased to notify you have met the 
> qualifications for 25k free OSS credits. To apply these credits to your 
> account, we will need your account's subscription to be on a Usage Based 
> subscription. Please select "Free Plan"[1] from your organizations 
> plan's page for a OSS only account and let us know when this step is 
> completed, or we can move your paid plan to a Usage Based[2] model if 
> your organization is using both OSS builds as well as paid builds.
> 
> Please let us know if you have any questions, we would be happy to assist.
> 
> Thanks so much and we will follow up on your reply.
> 
> 
> [1]- https://docs.travis-ci.com/user/billing-overview/#free-plan 
> <https://docs.travis-ci.com/user/billing-overview/#free-plan>
> [2]- https://docs.travis-ci.com/user/billing-overview/#usage-based-plans 
> <https://docs.travis-ci.com/user/billing-overview/#usage-based-plans>

My answer:
> Dear Travis CI team.
> 
> thank you for your reply and your offer.
> 
>> We are pleased to notify you have met the
>> qualifications for 25k free OSS credits.
> 
> On what frequency are these credits given out?
> 
> One build running to completion would be maybe 9hrs 
> (https://travis-ci.org/github/agda/agda/builds/765371923), that would be 
> say 5000 credits on Linux, so 25k last for 5 builds.
> 
> Best,
> Andreas

**Comment by andreasabel** [comment]
@asr
> At least the `hTags` test is missing in the GitHub workflows.

One could say that `make TAGS` is a developer-only feature and need not be tested with CI.  Or something akin to https://github.com/agda/agda/blob/5f947133322abe62215362d7ab3ac4a4d1928cb8/.github/workflows/whitespace.yml could be implemented.

**Comment by andreasabel** [comment]
I don't think a burning issue is left here.
